rs/src/main.rs

```diff
@@ -118,15 +118,24 @@ fn changed_components(components: Vec<&Path>, opts: &DiffOpts) -> Result<HashSet
         }
     }
     .into_iter()
-    .filter_map(|file_path| {
+    .flat_map(|file_path| {
         let file_path = Path::new(&file_path);
-        for component in &components {
-            if file_path.starts_with(component) {
-                return component.to_str().map(|s| s.to_owned());
-            }
-        }
 
-        None
+        components
+            .iter()
+            .filter_map(|component| {
+                if file_path.starts_with(component) {
+                    Some(
+                        component
+                            .to_str()
+                            .expect("a valid utf-8 component path")
+                            .to_owned(),
+                    )
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>()
```

Comment:
Probably more concise to do filter & map separately, you shouldn't need to collect _and_ I might be tempted to just assume `component` was utf8 since you read the path out of a UTF8 file in the first place (though that last bit is your choice obvs)

```suggestion
            .filter(|component| file_path.starts_with(component))
            .flat_map(Path::to_str)
            .map(ToOwned::to_owned)
```

alternatively if you want that expect:

```suggestion
            .filter(|component| file_path.starts_with(component))
            .map(|component| component.to_str().expect("a valid utf-8 component path").to_owned())
```

Reply:
Went for the first option, ta!